File: sdk/ts/src/receipt/rotation.test.ts
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import { verifyChain } from "./chain.js";
import { createReceipt } from "./create.js";
import { hashReceipt } from "./hash.js";
import {
	ed25519RawToPem,
	keyFingerprint,
	pemToEd25519Raw,
	verifyRotationEvent,
} from "./rotation.js";
import { generateKeyPair, signReceipt } from "./signing.js";
import type { AgentReceipt, KeyRotation } from "./types.js";

// RFC 8032 §7.1 well-known test public keys (raw 32-byte Ed25519), reused by
// the spec rotation-event vector. TEST 2 is the outgoing key (signs the
// rotation); TEST 3 is the incoming key.
const RFC8032_TEST2_PUB_HEX =
	"3d4017c3e843895a92b70aa74d1b7ebc9c982ccf2ec4968cc0cd55f12af4660c";
const RFC8032_TEST3_PUB_HEX =
	"fc51cd8e6218a1a38da47ed00230f0580816ed13ba3303ac5deb911548908025";

const here = dirname(fileURLToPath(import.meta.url));
const vectorPath = join(
	here,
	"..",
	"..",
	"..",
	"..",
	"spec",
	"test-vectors",
	"rotation-event",
	"example.json",
);

function pemFromHex(hex: string): string {
	return ed25519RawToPem(Buffer.from(hex, "hex"));
}

function loadVector(): AgentReceipt {
	return JSON.parse(readFileSync(vectorPath, "utf-8")) as AgentReceipt;
}

describe("key rotation", () => {
	it("verifies the spec rotation-event vector under the outgoing key", () => {
		const vector = loadVector();
		expect(vector.credentialSubject.keyRotation?.event_type).toBe(
			"key_rotated",
		);
		const outgoingPem = pemFromHex(RFC8032_TEST2_PUB_HEX);
		const cv = verifyChain([vector], outgoingPem);
		expect(cv.valid).toBe(true);
	});

	it("matches the published canonical-body hash", () => {
		const got = hashReceipt(loadVector());
		expect(got).toBe(
			"sha256:6983c9bd6fb24e844b90f7616315a914fdedc5fef8126e11d46149ba2f320457",
		);
	});

	it("binds the incoming key from the rotation event", () => {
		const vector = loadVector();
		const outgoingPem = pemFromHex(RFC8032_TEST2_PUB_HEX);
		const kr = vector.credentialSubject.keyRotation as KeyRotation;
		const rot = verifyRotationEvent(outgoingPem, kr);
		expect(rot.ok).toBe(true);
		if (rot.ok) {
			expect(rot.newKeyPem).toBe(pemFromHex(RFC8032_TEST3_PUB_HEX));
		}
	});

	it("rejects malformed rotation events", () => {
		const outgoingPem = pemFromHex(RFC8032_TEST2_PUB_HEX);
		const base = () =>
			structuredClone(
				loadVector().credentialSubject.keyRotation,
			) as KeyRotation;
		const zeroFp = `sha256:${"0".repeat(64)}`;
		const cases: Array<[string, (k: KeyRotation) => void, string]> = [
			[
				"bad event_type",
				(k) => ((k as { event_type: string }).event_type = "rotated"),
				"event_type",
			],
			[
				"bad signed_with",
				(k) => ((k as { signed_with: string }).signed_with = "new"),
				"signed_with",
			],
			[
				"unsupported old_algorithm",
				(k) => (k.old_algorithm = "ml-dsa"),
				"old_algorithm",
			],
			[
				"unsupported new_algorithm",
				(k) => (k.new_algorithm = "ml-dsa"),
				"new_algorithm",
			],
			[
				"old fingerprint mismatch",
				(k) => (k.old_key_fingerprint = zeroFp),
				"old_key_fingerprint",
			],
			[
				"new fingerprint mismatch",
				(k) => (k.new_key_fingerprint = zeroFp),
				"new_key_fingerprint",
			],
			[
				"new_public_key not multibase",
				(k) => (k.new_public_key = `z${k.new_public_key.slice(1)}`),
				"new_public_key",
			],
			[
				"new_public_key wrong length",
				(k) => (k.new_public_key = "uAAAA"),
				"new_public_key",
			],
		];
		for (const [, mutate, want] of cases) {
			const kr = base();
			mutate(kr);
			const rot = verifyRotationEvent(outgoingPem, kr);
			expect(rot.ok).toBe(false);
			if (!rot.ok) expect(rot.error).toContain(want);
		}
	});

	it("adopts the incoming key for receipts after a rotation", () => {
		const outKP = generateKeyPair();
		const inKP = generateKeyPair();
		const inRaw = pemToEd25519Raw(inKP.publicKey);

		const rot = createReceipt({
			issuer: { id: "did:agent:test" },
			principal: { id: "did:user:test" },
			action: { type: "agent.key.rotate", risk_level: "high" },
			outcome: { status: "success" },
			chain: {
				sequence: 1,
				previous_receipt_hash: null,
				chain_id: "chain_rot",
			},
		});
		rot.credentialSubject.keyRotation = {
			event_type: "key_rotated",
			new_public_key: `u${inRaw.toString("base64url")}`,
			old_key_fingerprint: keyFingerprint(pemToEd25519Raw(outKP.publicKey)),
			new_key_fingerprint: keyFingerprint(inRaw),
			old_algorithm: "ed25519",
			new_algorithm: "ed25519",
			signed_with: "old",
		};
		const signed0 = signReceipt(rot, outKP.privateKey, "did:agent:test#key-1");
		const h0 = hashReceipt(signed0);

		const r1 = createReceipt({
			issuer: { id: "did:agent:test" },
			principal: { id: "did:user:test" },
			action: { type: "filesystem.file.read", risk_level: "low" },
			outcome: { status: "success" },
			chain: { sequence: 2, previous_receipt_hash: h0, chain_id: "chain_rot" },
		});
		const signed1 = signReceipt(r1, inKP.privateKey, "did:agent:test#key-1");

		// Verified under only the outgoing genesis key — succeeds because the
		// rotation hands the key over.
		expect(verifyChain([signed0, signed1], outKP.publicKey).valid).toBe(true);

		// Without the rotation object, the successor signed by the incoming key
		// fails under the outgoing key alone. Omit keyRotation entirely (an
		// explicit undefined property would trip the canonicaliser).
		const { keyRotation: _omit, ...csNoRot } = rot.credentialSubject;
		const noRot = signReceipt(
			{ ...rot, credentialSubject: csNoRot },
			outKP.privateKey,
			"did:agent:test#key-1",
		);
		const h0b = hashReceipt(noRot);
		const r1b = createReceipt({
			issuer: { id: "did:agent:test" },
			principal: { id: "did:user:test" },
			action: { type: "filesystem.file.read", risk_level: "low" },
			outcome: { status: "success" },
			chain: { sequence: 2, previous_receipt_hash: h0b, chain_id: "chain_rot" },
		});
		const noRot1 = signReceipt(r1b, inKP.privateKey, "did:agent:test#key-1");
		expect(verifyChain([noRot, noRot1], outKP.publicKey).valid).toBe(false);
	});
});

File: sdk/go/receipt/rotation_test.go
package receipt

import (
	"crypto/ed25519"
	"crypto/x509"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"encoding/pem"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// RFC 8032 §7.1 well-known test public keys (raw 32-byte Ed25519), reused by
// the spec rotation-event vector. TEST 2 is the outgoing key (signs the
// rotation); TEST 3 is the incoming key.
const (
	rfc8032Test2PubHex = "3d4017c3e843895a92b70aa74d1b7ebc9c982ccf2ec4968cc0cd55f12af4660c"
	rfc8032Test3PubHex = "fc51cd8e6218a1a38da47ed00230f0580816ed13ba3303ac5deb911548908025"
)

func mustPEMFromHex(t *testing.T, hexKey string) string {
	t.Helper()
	raw, err := hex.DecodeString(hexKey)
	if err != nil {
		t.Fatalf("decode hex key: %v", err)
	}
	der, err := x509.MarshalPKIXPublicKey(ed25519.PublicKey(raw))
	if err != nil {
		t.Fatalf("marshal SPKI: %v", err)
	}
	return string(pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: der}))
}

func loadRotationVector(t *testing.T) AgentReceipt {
	t.Helper()
	path := filepath.Join("..", "..", "..", "spec", "test-vectors", "rotation-event", "example.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read rotation vector: %v", err)
	}
	var r AgentReceipt
	if err := json.Unmarshal(data, &r); err != nil {
		t.Fatalf("unmarshal rotation vector: %v", err)
	}
	return r
}

// TestRotationVectorVerifies pins the spec rotation-event vector: a
// genesis-position key_rotated receipt signed by the outgoing key must verify,
// and the parsed keyRotation must carry the seven fields.
func TestRotationVectorVerifies(t *testing.T) {
	r := loadRotationVector(t)
	if r.CredentialSubject.KeyRotation == nil {
		t.Fatal("rotation vector did not unmarshal a keyRotation object")
	}
	kr := r.CredentialSubject.KeyRotation
	if kr.EventType != "key_rotated" || kr.SignedWith != "old" {
		t.Fatalf("unexpected rotation constants: event_type=%q signed_with=%q", kr.EventType, kr.SignedWith)
	}

	outgoingPEM := mustPEMFromHex(t, rfc8032Test2PubHex)
	cv := VerifyChain([]AgentReceipt{r}, outgoingPEM)
	if !cv.Valid {
		t.Fatalf("rotation vector failed to verify: brokenAt=%d err=%q", cv.BrokenAt, cv.Error)
	}
}

// TestRotationVectorCanonicalHash cross-checks our canonicalization against the
// value the vector README publishes for the receipt body (proof removed).
func TestRotationVectorCanonicalHash(t *testing.T) {
	r := loadRotationVector(t)
	got, err := HashReceipt(r)
	if err != nil {
		t.Fatalf("hash receipt: %v", err)
	}
	const want = "sha256:6983c9bd6fb24e844b90f7616315a914fdedc5fef8126e11d46149ba2f320457"
	if got != want {
		t.Fatalf("canonical hash mismatch:\n got %s\nwant %s", got, want)
	}
}

// TestVerifyRotationEventBindsIncomingKey checks the field-level rotation
// validation and that the returned PEM is the incoming (TEST 3) key.
func TestVerifyRotationEventBindsIncomingKey(t *testing.T) {
	r := loadRotationVector(t)
	outgoingPEM := mustPEMFromHex(t, rfc8032Test2PubHex)

	newKeyPEM, err := verifyRotationEvent(outgoingPEM, r.CredentialSubject.KeyRotation)
	if err != nil {
		t.Fatalf("verifyRotationEvent: %v", err)
	}
	wantPEM := mustPEMFromHex(t, rfc8032Test3PubHex)
	if newKeyPEM != wantPEM {
		t.Fatalf("incoming key PEM mismatch:\n got %q\nwant %q", newKeyPEM, wantPEM)
	}
}

func TestVerifyRotationEventRejects(t *testing.T) {
	outgoingPEM := mustPEMFromHex(t, rfc8032Test2PubHex)
	base := func() KeyRotation { return *loadRotationVector(t).CredentialSubject.KeyRotation }

	cases := []struct {
		name   string
		mutate func(*KeyRotation)
		want   string
	}{
		{"bad event_type", func(k *KeyRotation) { k.EventType = "rotated" }, "event_type"},
		{"bad signed_with", func(k *KeyRotation) { k.SignedWith = "new" }, "signed_with"},
		{"unsupported old_algorithm", func(k *KeyRotation) { k.OldAlgorithm = "ml-dsa" }, "old_algorithm"},
		{"unsupported new_algorithm", func(k *KeyRotation) { k.NewAlgorithm = "ml-dsa" }, "new_algorithm"},
		{"old fingerprint mismatch", func(k *KeyRotation) { k.OldKeyFingerprint = "sha256:" + strings.Repeat("0", 64) }, "old_key_fingerprint"},
		{"new fingerprint mismatch", func(k *KeyRotation) { k.NewKeyFingerprint = "sha256:" + strings.Repeat("0", 64) }, "new_key_fingerprint"},
		{"new_public_key not multibase", func(k *KeyRotation) { k.NewPublicKey = "z" + k.NewPublicKey[1:] }, "new_public_key"},
		{"new_public_key wrong length", func(k *KeyRotation) { k.NewPublicKey = "uAAAA" }, "new_public_key"},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			kr := base()
			tc.mutate(&kr)
			_, err := verifyRotationEvent(outgoingPEM, &kr)
			if err == nil {
				t.Fatalf("expected error containing %q, got nil", tc.want)
			}
			if !strings.Contains(err.Error(), tc.want) {
				t.Fatalf("error %q does not mention %q", err.Error(), tc.want)
			}
		})
	}
}

func multibaseKey(raw []byte) string {
	return "u" + base64.RawURLEncoding.EncodeToString(raw)
}

// rotationChainReceipt0 builds the rotation receipt (seq 1) signed by the
// outgoing key, handing over to the incoming key.
func rotationChainReceipt0(t *testing.T, outKP, inKP KeyPair, withRotation bool) AgentReceipt {
	t.Helper()
	outRaw, _ := parsePublicKey(outKP.PublicKey)
	inRaw, _ := parsePublicKey(inKP.PublicKey)
	unsigned := Create(CreateInput{
		Issuer:    Issuer{ID: "did:agent:test"},
		Principal: Principal{ID: "did:user:test"},
		Action:    Action{Type: "agent.key.rotate", RiskLevel: RiskHigh},
		Outcome:   Outcome{Status: StatusSuccess},
		Chain:     Chain{Sequence: 1, PreviousReceiptHash: nil, ChainID: "chain_rot"},
	})
	if withRotation {
		unsigned.CredentialSubject.KeyRotation = &KeyRotation{
			EventType:         "key_rotated",
			NewPublicKey:      multibaseKey(inRaw),
			OldKeyFingerprint: keyFingerprint(outRaw),
			NewKeyFingerprint: keyFingerprint(inRaw),
			OldAlgorithm:      "ed25519",
			NewAlgorithm:      "ed25519",
			SignedWith:        "old",
		}
	}
	signed, err := Sign(unsigned, outKP.PrivateKey, "did:agent:test#key-1")
	if err != nil {
		t.Fatalf("sign rotation receipt: %v", err)
	}
	return signed
}

// TestRotationChainSwitchesKey proves the verifier adopts the incoming key for
// receipts after a rotation: a two-receipt chain whose successor is signed by
// the incoming key verifies under only the outgoing genesis key, precisely
// because the rotation receipt hands the key over.
func TestRotationChainSwitchesKey(t *testing.T) {
	outKP, err := GenerateKeyPair()
	if err != nil {
		t.Fatalf("gen outgoing: %v", err)
	}
	inKP, err := GenerateKeyPair()
	if err != nil {
		t.Fatalf("gen incoming: %v", err)
	}

	signed0 := rotationChainReceipt0(t, outKP, inKP, true)
	h0, err := HashReceipt(signed0)
	if err != nil {
		t.Fatalf("hash r0: %v", err)
	}
	r1 := Create(CreateInput{
		Issuer:    Issuer{ID: "did:agent:test"},
		Principal: Principal{ID: "did:user:test"},
		Action:    Action{Type: "filesystem.file.read", RiskLevel: RiskLow},
		Outcome:   Outcome{Status: StatusSuccess},
		Chain:     Chain{Sequence: 2, PreviousReceiptHash: &h0, ChainID: "chain_rot"},
	})
	signed1, err := Sign(r1, inKP.PrivateKey, "did:agent:test#key-1")
	if err != nil {
		t.Fatalf("sign r1: %v", err)
	}

	cv := VerifyChain([]AgentReceipt{signed0, signed1}, outKP.PublicKey)
	if !cv.Valid {
		t.Fatalf("rotation chain failed: brokenAt=%d err=%q", cv.BrokenAt, cv.Error)
	}

	// Without the rotation handover, the successor signed by the incoming key
	// must fail under the outgoing key alone.
	noRot0 := rotationChainReceipt0(t, outKP, inKP, false)
	h0b, _ := HashReceipt(noRot0)
	r1b := Create(CreateInput{
		Issuer:    Issuer{ID: "did:agent:test"},
		Principal: Principal{ID: "did:user:test"},
		Action:    Action{Type: "filesystem.file.read", RiskLevel: RiskLow},
		Outcome:   Outcome{Status: StatusSuccess},
		Chain:     Chain{Sequence: 2, PreviousReceiptHash: &h0b, ChainID: "chain_rot"},
	})
	noRot1, _ := Sign(r1b, inKP.PrivateKey, "did:agent:test#key-1")
	cv2 := VerifyChain([]AgentReceipt{noRot0, noRot1}, outKP.PublicKey)
	if cv2.Valid {
		t.Fatal("expected chain without rotation handover to fail successor signature, but it verified")
	}
}
